### PR TITLE
WIP radiasoft/container-jupyter-nvidia#9: compile hypre for gpu

### PR DESCRIPTION
**Pull request by e-carlin**
Same as rscode-elegant. Use the gpu-only flag to specify that gpu
specific compilation is needed.

Also, update hypre.

I tested the hypre change on opal and flash (the flash installer mentioned hypre). Both worked fine

### CODE REVIEW

File: installers/rpm-code/codes/hypre.sh [hunk 1]
```diff
@@ -2,9 +2,25 @@
 
 hypre_main() {
     codes_dependencies common
-    codes_download https://github.com/hypre-space/hypre/archive/v2.11.2.tar.gz
+    codes_download https://github.com/hypre-space/hypre/archive/v2.24.0.tar.gz
     cd src
-    ./configure --prefix="${codes_dir[prefix]}"
+    local a=''
```
Comment (robnagler) [inline installers/rpm-code/codes/hypre.sh:7]: `local -a a=()`

File: installers/rpm-code/codes/hypre.sh [hunk 2]
```diff
@@ -2,9 +2,25 @@
...
+    if [[ ${1:-} == 'gpu-only' ]]; then
+        #POSIT: container-jupyter-nvidia (cuda-11.1)
+        # and running on Tesla V100 (arch 70)
+        a=(
+            --with-mpi
+            --enable-unified-memory
+            --prefix="${codes_dir[prefix]}"
```
Comment (robnagler) [inline installers/rpm-code/codes/hypre.sh:14]: duplicate? it's passed below.

File: installers/rpm-code/codes/hypre.sh [hunk 3]
```diff
@@ -2,9 +2,25 @@
...
-    codes_download https://github.com/hypre-space/hypre/archive/v2.11.2.tar.gz
+    codes_download https://github.com/hypre-space/hypre/archive/v2.24.0.tar.gz
     cd src
-    ./configure --prefix="${codes_dir[prefix]}"
+    local a=''
+    if [[ ${1:-} == 'gpu-only' ]]; then
+        #POSIT: container-jupyter-nvidia (cuda-11.1)
+        # and running on Tesla V100 (arch 70)
+        a=(
+            --with-mpi
```
Comment (robnagler) [inline installers/rpm-code/codes/hypre.sh:12]: Is hypre compiled without mpi normally? Seems like we would want mpi.

File: installers/rpm-code/codes/hypre.sh [hunk 4]
```diff
@@ -2,9 +2,25 @@
...
+            --enable-unified-memory
+            --prefix="${codes_dir[prefix]}"
+            --with-MPI
```
Comment (robnagler) [inline installers/rpm-code/codes/hypre.sh:15]: is this a duplicate of with-mpi?

File: installers/rpm-code/codes/hypre.sh [hunk 5]
```diff
@@ -2,9 +2,25 @@
...
+            --with-cuda
+            --with-cuda-home='/usr/local/cuda-11.1'
```
Comment (robnagler) [inline installers/rpm-code/codes/hypre.sh:17]: You don't need quotes. Would be nice to get this from somewhere, but I know that's hard.

**Review by robnagler — CHANGES_REQUESTED** [review]

File: installers/rpm-code/codes/hypre.sh (hunk 2 above)
Comment (e-carlin) [reply installers/rpm-code/codes/hypre.sh]: 495c144

File: installers/rpm-code/codes/hypre.sh (hunk 5 above)
Comment (e-carlin) [reply installers/rpm-code/codes/hypre.sh]: 495c144

**Review by robnagler — APPROVED** [review]

**Pull request merged**